File: factories/Factory/FactoryMethod.cs
interface IProduct { }

class SmartPhone : IProduct { }

class Tablet : IProduct { }


abstract class ProductCreator{

    abstract IProduct FactoryMethod();
}

class  SmartphoneCreator : ProductCreator
{
    public override IProduct FactoryMethod(){
        return new SmartPhone();
    }
}

class TabletCreator : ProductCreator
{
    public override IProduct FactoryMethod(){
        return new Tablet();
    }
}

class Client{

    public void Main(){
        var smartPhone = SmartphoneCreator().FactoryMethod();
        var tablet = TabletCreator().FactoryMethod();
    }
}

File: factories/Factory/FactoryMethod.ts
namespace patterns.FactoryMethod {
  interface IProduct {}

  class SmartPhone implements IProduct {}
  class Tablet implements IProduct {}

  abstract class ProductCreator {
    abstract factoryMethod(): IProduct;
  }

  class SmartPhoneCreator extends ProductCreator {
    public factoryMethod(): IProduct {
      return new SmartPhone();
    }
  }

  class TabletCreator extends ProductCreator {
    public factoryMethod(): IProduct {
      return new Tablet();
    }
  }
}
